### PR TITLE
ESM2 Inference

### PR DESCRIPTION
**Pull request by farhadrgh**
(**NOTE:** _**delete** these instructional lines as you fill-out this PR template_)

(**NOTE:** _template is designed to be filled-in and used as the **squashed commit message for the entire PR**. _Italicized text_ is intended to be deleted as you fill in this template. Use the text between the `---`)

---

_High level summary of changes. Try to keep this as short and informative as possible: less is more._

_Describe your changes. You can be more detailed and descriptive here. If it is a code change, Be sure to answer:_
  - _What is changing?_
  - _What is the new or fixed functionality?_
  - _Why or when would someone want to use these changes?_
  - _How can someone use these changes?_
---

## Summary
_High level summary of changes. Try to keep this as short and informative as possible: less is more._

## Details
_Describe your changes. You can be more detailed and descriptive here._

## Usage
_How does a user interact with the changed code?_
```python
python -m your.new.module -and -all -options
```

## Testing
_How do you prove that your code behaves the way you claim?_

Tests for these changes can be run via:
```shell
pytest -v tests/your/new/or/existing/test_functions.py::test_function
```

(**NOTE:** _also **delete** this checklist as you fill-out this PR template_)

**Most of the changes** to files with extensions `*.py`, `*.yaml`, `*.yml`, `Dockerfile*` or `requirements.txt` **DO REQUIRE both `pytest-` and `jet-` CI stages**.

- [ ] Did you review the [Before your PR is "Ready for review" section](https://github.com/NVIDIA/bionemo-framework/-/blob/dev/CONTRIBUTING.md?ref_type=heads#before-pr-ready) before asking for review?
- [ ] Did you make sure your changes have tests? Did you test your changes locally?
- [ ] Can you add [the `SKIP_CI` label](https://github.com/NVIDIA/bionemo-framework/-/blob/dev/CONTRIBUTING.md?ref_type=heads#skip-ci) to your PR?
- [ ] Can you add [the `PYTEST_NOT_REQUIRED` label](https://github.com/NVIDIA/bionemo-framework/-/blob/dev/CONTRIBUTING.md?ref_type=heads#skip-pytest) to your PR?
- [ ] Can you add [the `JET_NOT_REQUIRED` label](https://github.com/NVIDIA/bionemo-framework/-/blob/dev/CONTRIBUTING.md?ref_type=heads#skip-jet) to your PR?

### CODE REVIEW

File: scripts/protein/esm2/esm2_infer.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    type=int,
+    required=False,
+    default=1,
+    help="Gradient accumulation steps. Global batch size is inferred from this.",
+)
+
+if __name__ == "__main__":
+    args = parser.parse_args()
+
+    results = infer_model(
```
Comment (jstjohn) [inline scripts/protein/esm2/esm2_infer.py:196]: `results = None`. Just call the function here, nothing is returned.

**Review by jstjohn — APPROVED** [review]

File: scripts/protein/esm2/esm2_infer.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+    parser.add_argument(
+        "--include-logits", action="store_true", default=False, help="Include per-token logits in output."
+    )
+    parser.add_argument(
+        "--accumulate-grad-batches",
+        type=int,
+        required=False,
+        default=1,
+        help="Gradient accumulation steps. Global batch size is inferred from this.",
+    )
```
Comment (skothenhill-nv) [inline scripts/protein/esm2/esm2_infer.py:249]: What does gradient accumulation do for inference? Can we remove this?

Comment (farhadrgh) [reply scripts/protein/esm2/esm2_infer.py:249]: hm, yeah good catch, will remove that

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/datamodule.py [hunk 3]
```diff
@@ -15,23 +15,114 @@
...
+
+    def __getitem__(self, index: int) -> BertSample:
+        """Obtains the BertSample at the given index."""
+        sequence = self.sequences[index]
+        tokenized_sequence = self._tokenize(sequence)
+
+        label = tokenized_sequence if len(self.labels) == 0 else self.labels[index]
+        # Overall mask for a token being masked in some capacity - either mask token, random token, or left as-is
+        loss_mask = ~torch.isin(tokenized_sequence, Tensor(self.tokenizer.all_special_ids))
+
```
Comment (skothenhill-nv) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/datamodule.py:80]: This gives me pause. We are using this Dataset for inference, but this is implying the tokenizer is doing some kind of masking/padding.

In the `ESM2FineTuneDataModule`, there are parameters for token masking and replacement, but it doesnt look like they are used. I think it would help to either document how these parameters are used (or might be used), or to remove them.  If I was using this for the first time, I would stare at the code for a while, set them to zero, and hope for the best. But I think we should either cover this in docs or not expose those parameters.

File: scripts/protein/esm2/esm2_infer.py [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+        pipeline_model_parallel_size=pipeline_model_parallel_size,
+        initial_ckpt_path=str(checkpoint_path),
+        initial_ckpt_skip_keys_with_these_prefixes=[],  # load everything from the checkpoint.
+    )
+
+    tokenizer = get_tokenizer()
+    module = biobert_lightning_module(config=config, tokenizer=tokenizer)
+
+    predictions = trainer.predict(module, datamodule=datamodule, return_predictions=True)
+    results_dict = batch_collator(predictions)
```
[the remark this answers is not in the document]
Comment (farhadrgh) [reply scripts/protein/esm2/esm2_infer.py:135]: or maybe I can just directly pass trainer.predict into the collator?

**Review by skothenhill-nv — APPROVED** [review]
Looks good, but could use some clarity on whats happening with masking/random tokens between the tokenizer, dataset, and datamodule.

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/datamodule.py (hunk 3 above)
Comment (farhadrgh) [reply sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/datamodule.py:81]: Yeah, I agree, I forgot to remove the masking parameters that is used for training datamodule and not here. I removed them in [f6dce08](https://github.com/NVIDIA/bionemo-framework/pull/384/commits/f6dce083fb092a12a1c001dfaeb92e3bc5895055)

File: sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/datamodule.py [hunk 5]
```diff
@@ -15,23 +15,118 @@
...
 from bionemo.esm2.model.finetune.finetune_regressor import InMemorySingleValueDataset
 from bionemo.esm2.model.finetune.finetune_token_classifier import InMemoryPerTokenValueDataset
 from bionemo.llm.data import collate
+from bionemo.llm.data.datamodule import MegatronDataModule
+from bionemo.llm.data.types import BertSample
 from bionemo.llm.utils.datamodule_utils import infer_num_samples
 
 
-class ESM2FineTuneDataModule(pl.LightningDataModule):
+class InMemoryCSVDataset(Dataset):
```
Comment (pstjohn) [inline sub-packages/bionemo-esm2/src/bionemo/esm2/model/finetune/datamodule.py:41]: I'd rather you separate the file IO from the dataset logic; especially if this is all just in-memory anyways.

Maybe just `ProteinDataset` or `ESMProteinDataset` that takes in `sequences: pd.Series, labels: pd.Series` or something like that. Then you could have a small convenience function outside the class (or a `@classmethod` alternative constructor) that creates this from a file path specifically.

**Review by pstjohn — APPROVED** [review]
